### DISCUSSION TITLE
Notification buttons that trigger actions

### DISCUSSION THREAD

**Discussion by thaije** [zauberzeug/nicegui#1994] [Q&A] [answered]
### Question

I'm triying to figure out how to trigger an action other than close when clicking a notification button. E.g. with as goal something like below, a notification with a button that opens the corresponding detection / error / etc.

![image](https://github.com/zauberzeug/nicegui/assets/2978372/d2000e7e-5701-42e3-ae4c-0ae695ea5e92)

To do so I followed the Quasar documentation that specifies how to use [`actions`](https://quasar.dev/quasar-plugins/notify#with-actions) for notifications. Passing that param as below works for me in NiceGUI, except that it throws an `Type is not JSON serializable: function` error about the handler function. 

A minimum example is: 
```python
from nicegui import ui

def test():
    print("hello")

def open_notification(): 
    ui.notify("Notification",
        actions=[
            { "label": 'View', "color": 'white', "handler": test'}
        ],
    )

ui.button("Trigger notification", on_click=open_notification)
ui.run()
```

So I don't understand how to pass a function that handles the button click of a notification. Does anyone have any tips?

Some other things I tried: 

## Lambda function as handler
```python
ui.notify("Notification",
    actions=[
        { "label": 'View', "color": 'white', "handler": None}
    ],
).on('click', lambda: ui.notify('You clicked the button B.'))
 ```
This also throws a `Type is not JSON serializable: function` error.

## Specify on.click function for notification
```python
ui.notify("Notification",
    actions=[
        { "label": 'View', "color": 'white', "handler": None}
    ],
).on('click', lambda: ui.notify('You clicked the button.'))
```
This throws an `AttributeError: 'NoneType' object has no attribute 'on'`.

## Run a javascript function on button click 
```python
ui.notify("Notification",
    actions=[
        { "label": 'View', "color": 'white', "handler": 'alert("hello")'}
    ],
)
```
This just does nothing, most likely as the passed handler function isn't parsed as javascript.

**Comment by falkoschindler** [comment, marked as answer]
Hi @thaije, unfortunately `ui.notify` is currently a bit limited. As soon as the notification is sent, the backend has no control over it anymore. So we need to get creative:

```py
ui.button('Notify', on_click=lambda e: ui.run_javascript('''
    Quasar.Notify.create({
        message: 'Hello!',
        actions: [{
            label: 'Alert',
            handler: () => getElement(''' + str(e.sender.id) + ''').$emit('alert_clicked'),
        }],
    })
''')).on('alert_clicked', lambda: print('Alert clicked!'))
```

Here we call Quasar's Notify API directly using `run_javascript`. The "Alert" action defines a handler which emits a custom event on the button element. This is caught using the generic event registration `.on()`. I hope this helps.

As a side note, here is a similar discussion about interaction with `ui.notify`: #515. Maybe we'll find a neat API for such things in the future.

**Reply by thaije** [reply]
Hey @falkoschindler, thanks for the quick reply! That is indeed a neat workaround.

**Comment by newrandom** [comment]
I unintentionally encountered a console error with the following message:
`Failed to execute 'setAttribute' on 'Element': 'handler: () => {console.log("test")}' is not a valid attribute name.
    at lo (vue.global.prod.js:5:64358)`

From this error, I confirmed that the setAttribute setting was incorrect.
So, I tried setting 'onclick': 'console.log("Hello")'.

As a result, I was able to see the console log when clicking the notify button.

Here is the code I wrote:
`ui.notify('TEST', type='positive', actions=[{'icon': 'refresh', 'label': '새로고침', 'onclick': 'console.log("Hello")'}])`

When clicking the button, I was able to see the following result in the console.
<img width="917" alt="image" src="https://github.com/user-attachments/assets/d170900b-288e-41a3-9abc-b7c77dc86b3f" />

and 
<img width="701" alt="image" src="https://github.com/user-attachments/assets/ebbea1fb-98cb-4062-849a-5cd0d48bb8ac" />

**Comment by defkev** [comment]
I know I am late to this but since nicegui supports “generic events” you can just use `emitEvent` on the client:

```py
from nicegui import ui

ui.notification(
   message = 'foo',
   caption = 'bar',
   type = 'warning',
   actions = [
      {'label': 'View', ':handler': 'emitEvent("view")'} # mind the : denoting emitEvent is Javascript
   ]
)
ui.on('view', lambda: print('Hello!'))
```
Which makes it much more Pythonic

Cheers

EDIT
Seemingly, this only works with `notification` and not `notify` ... odd.

**Comment by evnchn** [comment]
Was browsing through old notifications. Realized #4819 is highly related to this discussion.
